Objects/regions.c

```diff
@@ -92,9 +116,324 @@ struct regionmetadata {
+        return self;
+    }
+
+    // FIXME: It can happen that there are several layers in this union-find
+    // structure. It would be efficient to directly update the parent pointers
+    // for deeper nodes.
+    return regionmetadata_get_root(Py_region_ptr(self_data->parent));
+}
+#define regionmetadata_get_root(self) \
+    regionmetadata_get_root(_Py_CAST(Py_region_ptr_t, self))
+
+// This returns the represenative of this union-find structure or NULL
+// if the representative is a constant value.
+static regionmetadata* regionmetadata_get_root_or_null(Py_region_ptr_t self)
+{
+    Py_region_ptr_t rep = regionmetadata_get_root(self);
+    // Test for constants
+    if (IS_CONST_REGION(rep)) {
+        return NULL;
+    }
+
+    return _Py_CAST(regionmetadata*, rep);
+}
+#define regionmetadata_get_root_or_null(self) \
+    regionmetadata_get_root_or_null(_Py_CAST(Py_region_ptr_t, (self)))
+
+static void regionmetadata_open(regionmetadata* self) {
+    self = regionmetadata_get_root_or_null(self);
+    if (self) {
+        self->is_open = true;
+    }
+}
+
+static bool regionmetadata_is_open(regionmetadata* self) {
+    self = regionmetadata_get_root_or_null(self);
+    if (!self) {
+        // All constant regions are open
+        return true;
+    }
+
+    return self->is_open;
+}
+
+static void regionmetadata_inc_rc(Py_region_ptr_t self_raw)
+{
+    // Constant regions don't keep track of open subregions.
+    if (IS_REGION_NULL(self_raw) || IS_CONST_REGION(self_raw)) {
+        return;
+    }
+
+    regionmetadata* self = _Py_CAST(regionmetadata*, self_raw);
+
+    // This doesn't use the representative, since we want
+    // to track the RC on each individual region metadata object.
+    self->rc += 1;
+}
+#define regionmetadata_inc_rc(self) \
+    (regionmetadata_inc_rc(_Py_CAST(Py_region_ptr_t, self)))
+
+static void regionmetadata_dec_rc(Py_region_ptr_t self_raw)
+{
+    // Constant regions don't keep track of open subregions.
+    if (IS_REGION_NULL(self_raw) || IS_CONST_REGION(self_raw)) {
+        return;
+    }
+
+    regionmetadata* self = _Py_CAST(regionmetadata*, self_raw);
+
+    // This doesn't use the representative, since we want
+    // to track the RC on each individual region metadata object.
+    self->rc -= 1;
+    if (self->rc == 0) {
+        // This access the parent directly to update the rc.
+        // It also doesn't matter if the parent pointer is a
+        // merge or subregion relation, since both cases have
+        // increased the rc.
+        regionmetadata_dec_rc(Py_region_ptr(self->parent));
+
+        free(self);
+    }
+}
+#define regionmetadata_dec_rc(self) \
+    (regionmetadata_dec_rc(_Py_CAST(Py_region_ptr_t, self)))
+
+static void regionmetadata_inc_osc(Py_region_ptr_t self_raw)
+{
+    self_raw = regionmetadata_get_root(self_raw);
+
+    // Constant regions don't keep track of open subregions.
+    if (IS_CONST_REGION(self_raw)) {
```

Comment:
This is just a note that it is not clear to me what the meaning of `_raw` is and I don't see it documented anywhere. If it means an untagged pointer, maybe have an assert to document that here which would be compiled away?

Reply:
`_raw` was meant to indicate that this is the pointer or `Py_LOCAL` or `Py_IMMUTABLE`. This is also expressed in the type. That's why these functions take `Py_region_ptr_t` and not `regionmetadata*`. I've now renamed them to `self_prt`